### DISCUSSION TITLE
Livewire Validation - Be able to have the values entered in the fields be displayed if the validation fails, rather then be blanked out.

### DISCUSSION THREAD

**Discussion by lcallif** [livewire/livewire#9017] [7. Bugs (archived)]
Currently, if the Validation fails the User is taken back to the form, displaying the error messages (if any) that the developer has enabled, and clears all the fields where wire:model is used.  Not User friendly, not even close.

I propose that Livewire be enhanced to allow for the old values to be brought back and displayed. Yep, I know Laravel's "old" is not available to resolve this and understand the reasoning behind it.

Maybe we could use something like:
1) The following
```
$this->validate([
'addresstype' => 'required|gt:0',
'address' => 'required|min:2',])->keepVals();

```
And/or 

2) Allow for the specific fields to retain the original values

```
$this->validate([
        'addresstype' => 'required|gt:0',
        'address' => 'required|min:2',])->keepVals([addresstype, address]);
```

And/or
3) OR document a process for us to get the desired result(s).....

The current way, losing the values and forcing the User to re-enter ALL of the values is not user friendly. It is ticking off our Customers & our management and we are almost at the point where we will be forced to remove the use of Livewire unless we come up with a viable solution. And given that we are now instituting Flux, we really do not want to give up on Livewire.

And before somebody beats on me for specifying the field names in the validate clause, there is a good reason behind doing this. Most of the forms all have multiple parts:

The Subject Matter - assume Customer, allows for the entry of multiple sub-areas, all on one form:
0 or more addresses
0 or more emails
0 or more phones
0 or more background checks
some others
Without specifying the fields when executing the validation for the specific sub-area we would execute all of the fields for all of the sub-areas (addresses, emails, etc). Erroneous errors (say that fast) would be displayed. Definitely not User friendly.....

Suggestions to resolve this issue, or a timeline when this might be addressed by any of the 3 potential solutions I gave above, or a better solution than I gave.....

**Comment by joshhanley** [comment]
@lcallif Livewire's failed validation should not clear the values. It doesn't for me. Sounds like you have something else going on. Can I suggest creating a fresh Laravel app and testing validation in a basic scenario just to confirm that works as you are expecting? If that then fails, can you share the component class and view for that test here so we can have a look? Thanks!

**Reply by lcallif** [reply]
I will work on this and get back in a few days.  On to adding another Discussion entry related to validation.

**Reply by lcallif** [reply]
Hey Josh, would it be possible for you to post your code, please ?  It will be MUCH easier for me to see what I am doing wrong and what I need to change.

Please ?

**Comment by lcallif** [comment]
I decided that I would rewatch the Livewire's Validation video to see if I could glean any nuggets of information.  I saw that there were 2 things mentioned/shown that I was not using: 
```
#[Validation]
public $address;
```

And a section for the public function rules.  

```
    public function rules(){
        return [
            'address' => "required|min:2",
        ];
    }
```

Once I added something in for the above 2 sections, the values stayed in the fields that had validation issues.  While I don't understand why this solved the issue, I am not sure if this is a bug or may need to be highlighted or noted in the documentation.  I think it might be a bug, but then again I could be skirting a Livewire requirement.

I am voting for it being a bug because as long as I have 1 field noted as ``` #[Validate]``` and in a rules function, I no longer have an issue with ANY of the fields being validated.

Note, the following is the current code I am using for validating the address type and address & email type and email.  Just in case you need or want it.  Maybe something I did is what was causing the original issue.

```
    public function addAddress($id)
    {
        $validated = $this->validate([
            'addresstype' => 'required|gt:0',
            'address' => 'required|min:2',
        ]);
        $this->subcomponentinsertupdateform->addGeneric2Fields($id, $this->theAreaSingular . 'addresses',
            $this->theAreaSingular . '_id', $this->addresstype, $this->address, 'address');
        $this->oldAddress='';
        $this->oldAddresstype='';
        return redirect($this->redirectUrl);
    }

    public function addEmail($id)
    {
        $validated = $this->validate([
            'emailtype' => 'required|gt:0',
            'email' => 'required|email',]);
        $this->subcomponentinsertupdateform->addGeneric2Fields($id, $this->theAreaSingular . 'emails',
            $this->theAreaSingular . '_id', $this->emailtype, $this->email, 'email');
        $this->emails = CustomerEmail::where('customer_id',$this->theID)
            ->get();
        return redirect($this->redirectUrl);
    }

```

Yeah I know, I should probably use the model's name to create the entry, but I have a realllllly good reason for this approach.
```
class SubComponentInsertUpdateForm extends Form
{
    public function addGeneric2Fields($id, $tableName, $area_Id, $theType, $theVal, $specificFieldName)
    {
        DB::table($tableName)->insert([
            'id' => Str::uuid(),
            'user_id' => Auth::id(),
            'team_id' => Auth::user()->currentTeam->id,
            $area_Id => $id,
            'typeId' => $theType,
            $specificFieldName => $theVal,
            'created_at' => now(),
        ]);
    }
}
```

There are more messages being used in the Messages function, but there wasn't a reason to include them. 
```
    public function messages()
    {
        return [

            'addresstype' => 'Address Type is Required',
            'address'     => 'Address - Minimum of 2 Characters',
            'addressInputs.*.address' => 'Address - Minimum of 2 Characters',

            'emailtype' => 'Email Type is Required',
            'email.required' => 'Email - Minimum of 5 Characters',
            'email.email' => 'Invalid Email',
            'emailInputs.*.email' => 'Invalid Email',

        ];
    }
```

Thank you for your time.

**Comment by lcallif** [comment]
Update, this is now a bonfide Bug.

The ONLY thing needed is a ```#[Validated]``` for anything, even for something that is not used in this component.  So something like the following without a rules function allows the error field to retain their value.

```
    #[Validate]
    public $junkerrrrr;
```

Again, no rules function.
